jwst/undersampling_correction/undersampling_correction.py

```diff
@@ -87,18 +91,50 @@ def flag_pixels(data, gdq, signal_threshold):
             if len(wh_exc_1d[0] > 0):  # For ramps previously unflagged ...
                 lowest_exc_1d[wh_exc_1d] = ii_grp
 
-    # Flag current and subsequent groups
+    # Loop over exceedances and flag as UNDERSAMP and DO_NOT_USE, then flag
+    #   that pixel's 4 nearest neighbors
     lowest_exc_2d = lowest_exc_1d.reshape((n_rows, n_cols))
+
     for ii_int in range(n_ints):
         for ii_grp in range(n_grps):
             wh_set_flag = np.where(lowest_exc_2d == ii_grp)
-
-            # set arrays of components
+            # set arrays of coordinates of each exceedance
             yy = wh_set_flag[0]
             xx = wh_set_flag[1]
 
-            gdq[ii_int, ii_grp:, yy, xx] = \
-                np.bitwise_or(gdq[ii_int, ii_grp:, yy, xx], dqflags.group['UNDERSAMP']
-                              | dqflags.group['DO_NOT_USE'])
+            if len(wh_set_flag) > 0 and len(xx) > 0:  # there are exceedances
+                gdq[ii_int, ii_grp:, yy, xx] = \
+                    np.bitwise_or(gdq[ii_int, ii_grp:, yy, xx],
+                                  dqflags.group['UNDERSAMP']
+                                  | dqflags.group['DO_NOT_USE'])
+
+                # Set the same flags for the 4 nearest neighbors
+                wh_west = np.where(xx > 0)  # pixels whose western neighbors to be flagged
+                if len(wh_west[0]) > 0:
```

Comment:
Given that a bitwise_or is being used in all of these operations and hence there's no danger of flagging a pixel twice with the same flag, I wonder if these four operations (one for each neighboring pixel) could be simplified to be done in only 2 operations - one for horizontal ("east", "west") neighbors and one for vertical ("north", "south") neighbors. Something like:

np.bitwise_or(gdq[ii_int, ii_grp, yy, xx-1:xx+1], dqflags.group ...)    # for the 3 horizontal pixels
np.bitwise_or(gdq[ii_int, ii_grp, yy-1:yy+1, xx], dqflags.group ...)    # for the 3 vertical pixels

Yes, the center (original) pixel gets operated on multiple times, but it doesn't hurt anything (because of the "or") and it greatly simplifies the code.

Reply:
I couldn't refactor this way, but I did improve the efficiency by replacing calls of np.where() using arrays.